### ISSUE TITLE
Reproduction of the key incorrect

### ISSUE THREAD

**Issue by Plaenkler** [pquerna/otp#83]
I am facing the challenge to generate the QR code again after restarting my application. The only thing that is permanently saved is the key.secret. However, it does not seem to be possible to recover the same key from the secret only.

```golang
// This code section made me think that I can recover the key only from the secret.
// totp.go 182-192
v := url.Values{}
if len(opts.Secret) != 0 {
    v.Set("secret", b32NoPadding.EncodeToString(opts.Secret))
} else {
    secret := make([]byte, opts.SecretSize)
    _, err := opts.Rand.Read(secret)
    if err != nil {
        return nil, err
    }
    v.Set("secret", b32NoPadding.EncodeToString(secret))
}
```

However, my latest tests seem to disprove this theory.

```golang
fmt.Println(keySecret)
key, err := totp.Generate(totp.GenerateOpts{
    Issuer:      issuer,
    AccountName: accountName,
    Secret:      []byte(keySecret),
})
fmt.Println(key.Secret())
```

keySecret: JRKDOQRSJJBDIVCRLJJEQS22KBGQINGFQMRUGZHDGNKYINBUSRKI
key.Secret(): KKKJFUIT2RKJJUUSSCIREVMQ2SJRFEURKRKMZDES2CI5IQJFHEORSRJVJFKR22JBCEOTSLLFEU4QSVKNJEWS

The issue should be reproducible with any given key

**Comment by pquerna** [comment]
I'm not sure this needs a new function, named Regenerate, but rather more clarity about the code flows for serialization and deserialization.

**Comment by Plaenkler** [comment]
> I'm not sure this needs a new function, named Regenerate, but rather more clarity about the code flows for serialization and deserialization.

Of course, a note in the documentation and an example could help here. However, I think the user should not be responsible for the decryption of the secret. Basically, as a user, I would assume that the set option "Secret", which is also a field of the key is used unprocessed.  The function does not introduce any breaking changes, which is why I would prefer it to a correction in the documentation.

**Comment by pquerna** [comment]
Need to be clear here, its not "encrypted";  Its base32 encoded.

**Comment by Plaenkler** [comment]
I had made a mistake in my choice of words. Regardless of whether it is encoded or encrypted, the library should provide a way to recover the key from the secret.

**Comment by DavesPlanet** [comment]
Plaenkler, thanks for saving my hide with your example! I would never have figured that out.
One minor tweek, the code does not seem to run without allocating space for the buffer

var secret []byte

needed to become this for me to get it to work

secret := make([]byte, base32.StdEncoding.DecodedLen(len(opts.Secret)))

**Comment by Plaenkler** [comment]
> Plaenkler, thanks for saving my hide with your example! I would never have figured that out. One minor tweek, the code does not seem to run without allocating space for the buffer
> 
> var secret []byte
> 
> needed to become this for me to get it to work
> 
> secret := make([]byte, base32.StdEncoding.DecodedLen(len(opts.Secret)))

Thanks for the info have adjusted it in the pull request
